File: jobs/2026-05-27__02-20-35/llamacloud_extract_with_citation__HWKBYan/artifacts/user/llamacloud-task/extract_invoice.py
import json
import os
import time
from pathlib import Path

from llama_cloud import BadRequestError, LlamaCloud
from pydantic import BaseModel, Field


class InvoiceSchema(BaseModel):
    company_name: str = Field(..., description="Name of the company issuing the invoice")
    invoice_number: str = Field(..., description="Invoice identifier")
    total_amount: float = Field(..., description="Total amount due on the invoice")


def main() -> None:
    run_id = os.environ.get("ZEALT_RUN_ID")
    if not run_id:
        raise RuntimeError("ZEALT_RUN_ID environment variable is required")

    client = LlamaCloud()

    invoice_path = Path(__file__).resolve().parent / "sample_invoice.txt"
    if not invoice_path.exists():
        raise FileNotFoundError(f"Missing invoice file: {invoice_path}")

    external_file_id = f"invoice-{run_id}.txt"
    try:
        uploaded_file = client.files.create(
            file=str(invoice_path),
            purpose="extract",
            external_file_id=external_file_id,
        )
    except BadRequestError as exc:
        if "duplicate key" not in str(exc):
            raise
        existing_files = list(client.files.list(external_file_id=external_file_id))
        if not existing_files:
            raise
        uploaded_file = existing_files[0]

    extraction_job = client.extract.create(
        file_input=uploaded_file.id,
        configuration={
            "data_schema": InvoiceSchema.model_json_schema(),
            "extraction_target": "per_doc",
            "tier": "agentic",
            "cite_sources": True,
        },
    )

    terminal_statuses = {"COMPLETED", "FAILED", "CANCELLED"}
    while extraction_job.status not in terminal_statuses:
        time.sleep(2)
        extraction_job = client.extract.get(
            extraction_job.id, expand=["extract_metadata"]
        )

    extraction_job = client.extract.get(
        extraction_job.id, expand=["extract_metadata"]
    )

    log_path = invoice_path.parent / "output.log"
    log_path.write_text(
        f"Extract job: {extraction_job.id}\nStatus: {extraction_job.status}\n",
        encoding="utf-8",
    )

    if extraction_job.status != "COMPLETED":
        raise RuntimeError(f"Extraction failed with status: {extraction_job.status}")

    extract_result = extraction_job.extract_result
    if hasattr(extract_result, "model_dump"):
        extract_payload = extract_result.model_dump(mode="json")
    else:
        extract_payload = extract_result

    extract_metadata = extraction_job.extract_metadata
    if hasattr(extract_metadata, "model_dump"):
        extract_metadata_payload = extract_metadata.model_dump(mode="json")
    else:
        extract_metadata_payload = extract_metadata or {}

    field_metadata = {}
    if isinstance(extract_metadata_payload, dict):
        field_metadata = extract_metadata_payload.get("field_metadata", {})
        if isinstance(field_metadata, dict) and "document_metadata" in field_metadata:
            field_metadata = field_metadata.get("document_metadata", {})

    normalized_extract_metadata = {"field_metadata": field_metadata}
    if isinstance(extract_metadata_payload, dict):
        if "parse_job_id" in extract_metadata_payload:
            normalized_extract_metadata["parse_job_id"] = extract_metadata_payload[
                "parse_job_id"
            ]
        if "parse_tier" in extract_metadata_payload:
            normalized_extract_metadata["parse_tier"] = extract_metadata_payload[
                "parse_tier"
            ]

    output_payload = {
        "data": extract_payload,
        "extract_metadata": normalized_extract_metadata,
    }

    result_path = invoice_path.parent / "result.json"
    result_path.write_text(json.dumps(output_payload, indent=2), encoding="utf-8")


if __name__ == "__main__":
    main()

File: jobs/2026-07-03__05-39-37/llamacloud_extract_invoice_field__ZW7RExV/artifacts/user/project/extract-invoice.ts
/**
 * Extract structured invoice data from a PDF using LlamaCloud Extract v2.
 *
 * Uses the official @llamaindex/llama-cloud SDK (v2) with a Zod-based schema
 * converted to JSON Schema via z.toJSONSchema(...).
 *
 * Usage:  npx tsx extract-invoice.ts
 */

import fs from "node:fs";
import path from "node:path";
import { LlamaCloud, toFile } from "@llamaindex/llama-cloud";
import { z } from "zod";

const INVOICE_PDF = path.join(__dirname, "invoice.pdf");
const OUTPUT_JSON = path.join(__dirname, "output.json");
const OUTPUT_LOG = path.join(__dirname, "output.log");

// ---------------------------------------------------------------------------
// 1. Define the invoice schema with Zod
// ---------------------------------------------------------------------------
const lineItemSchema = z.object({
  description: z.string().describe("Description of the line item / product / service"),
  quantity: z.number().describe("Quantity ordered"),
  unit_price: z.number().describe("Price per unit"),
  total: z.number().describe("Total for this line item (quantity * unit_price)"),
});

const invoiceSchema = z.object({
  invoice_number: z.string().describe("The unique invoice number / identifier"),
  invoice_date: z
    .string()
    .describe("Invoice date in ISO YYYY-MM-DD format if possible"),
  vendor_name: z.string().describe("Name of the vendor / supplier issuing the invoice"),
  total_amount: z
    .number()
    .describe("The grand total amount of the invoice in the invoice's currency"),
  line_items: z.array(lineItemSchema).describe("Itemised list of charges on the invoice"),
});

// Convert the Zod schema into a JSON Schema object accepted by the API.
const dataSchema = z.toJSONSchema(invoiceSchema) as Record<string, unknown>;

// ---------------------------------------------------------------------------
// 2. LlamaCloud client (reads LLAMA_CLOUD_API_KEY from the environment)
// ---------------------------------------------------------------------------
const client = new LlamaCloud();

// ---------------------------------------------------------------------------
// Helper: poll an extract job until it reaches a terminal state
// ---------------------------------------------------------------------------
const TERMINAL_STATES = new Set(["COMPLETED", "FAILED", "CANCELLED"]);

async function pollUntilTerminal(jobId: string) {
  const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
  let job = await client.extract.get(jobId);

  while (!TERMINAL_STATES.has(job.status)) {
    process.stdout.write(
      `  ...job ${jobId} status: ${job.status}. Waiting 5s...\n`
    );
    await sleep(5000);
    job = await client.extract.get(jobId);
  }

  return job;
}

// ---------------------------------------------------------------------------
// 3. Main workflow
// ---------------------------------------------------------------------------
async function main() {
  console.log("=== LlamaCloud Invoice Extractor ===");

  // --- Upload the invoice PDF (purpose: "extract") -------------------------
  console.log(`Uploading ${INVOICE_PDF} ...`);
  const fileBlob = await toFile(fs.readFileSync(INVOICE_PDF), "invoice.pdf", {
    type: "application/pdf",
  });
  const uploaded = await client.files.create({
    file: fileBlob,
    purpose: "extract",
  });
  const fileId = uploaded.id;
  console.log(`  Uploaded file id: ${fileId}`);

  // --- Create the Extract v2 job (agentic tier, per_doc) ------------------
  console.log("Creating Extract v2 job (tier: agentic, extraction_target: per_doc)...");
  const job = await client.extract.create({
    file_input: fileId,
    configuration: {
      data_schema: dataSchema,
      extraction_target: "per_doc",
      tier: "agentic",
    },
  });
  const jobId = job.id;
  console.log(`  Created job id: ${jobId} (initial status: ${job.status})`);

  // --- Poll until terminal ------------------------------------------------
  console.log("Polling for completion...");
  const completed = await pollUntilTerminal(jobId);
  console.log(`  Final status: ${completed.status}`);

  if (completed.status !== "COMPLETED") {
    const errMsg = completed.error_message ?? "(no error message provided)";
    throw new Error(`Extract job ${completed.status}: ${errMsg}`);
  }

  // --- Write the parsed extract_result to output.json ---------------------
  const extractResult = completed.extract_result;
  const jsonText = JSON.stringify(extractResult, null, 2);
  fs.writeFileSync(OUTPUT_JSON, jsonText + "\n", "utf-8");
  console.log(`Wrote parsed result to ${OUTPUT_JSON}`);

  // --- Build & append the one-line summary to output.log ------------------
  // extract_result is a single object for per_doc extraction.
  const resultObj = (Array.isArray(extractResult) ? extractResult[0] : extractResult) as
    | Record<string, unknown>
    | null;

  const invoiceNumber = (resultObj?.invoice_number as string) ?? "UNKNOWN";
  const vendorName = (resultObj?.vendor_name as string) ?? "UNKNOWN";
  const totalAmount = (resultObj?.total_amount as number) ?? "UNKNOWN";

  const summaryLine = `Extracted Invoice: ${invoiceNumber} | Vendor: ${vendorName} | Total: ${totalAmount}`;

  // Idempotent: overwrite output.log on every run.
  fs.writeFileSync(OUTPUT_LOG, summaryLine + "\n", "utf-8");
  console.log(`Wrote summary to ${OUTPUT_LOG}:`);
  console.log(`  ${summaryLine}`);
}

main().catch((err) => {
  console.error("Extract failed:", err);
  process.exit(1);
});